docker/bookkeeper/entrypoint.sh

```diff
@@ -8,68 +8,38 @@
-echo "BK_LEDGERS_PATH is $BK_LEDGERS_PATH"
-
-sed -i "s|bookiePort=.*\$|bookiePort=${PORT0}|" ${BK_HOME}/conf/bk_server.conf
-sed -i "s|metadataServiceUri=.*\$|metadataServiceUri=zk://${ZK_URL}${BK_LEDGERS_PATH}|" ${BK_HOME}/conf/bk_server.conf
-sed -i "/zkServers=.*/d" ${BK_HOME}/conf/bk_server.conf
-sed -i "s|journalDirectory=.*\$|journalDirectory=${BK_DIR}/journal|" ${BK_HOME}/conf/bk_server.conf
-sed -i "s|ledgerDirectories=.*\$|ledgerDirectories=${BK_DIR}/ledgers|" ${BK_HOME}/conf/bk_server.conf
-sed -i "s|indexDirectories=.*\$|indexDirectories=${BK_DIR}/index|" ${BK_HOME}/conf/bk_server.conf
-sed -i "s|# autoRecoveryDaemonEnabled=.*\$|autoRecoveryDaemonEnabled=${BK_AUTORECOVERY}|" ${BK_HOME}/conf/bk_server.conf
-sed -i "s|# useHostNameAsBookieID=.*\$|useHostNameAsBookieID=${BK_useHostNameAsBookieID}|" ${BK_HOME}/conf/bk_server.conf
-
-echo "
-tlsProvider=OpenSSL
-# key store
-tlsKeyStoreType=JKS
-tlsKeyStore=/var/private/tls/bookie.keystore.jks
-tlsKeyStorePasswordPath=/var/private/tls/bookie.keystore.passwd
-# trust store
-tlsTrustStoreType=JKS
-tlsTrustStore=/var/private/tls/bookie.truststore.jks
-tlsTrustStorePasswordPath=/var/private/tls/bookie.truststore.passwd" >> ${BK_HOME}/conf/bk_server.conf
+BK_DIR="/bk"
+
+export BK_zkServers=$ZK_URL
+export BK_metadataServiceUri=zk://$ZK_URL$BK_LEDGERS_PATH
+export BK_journalDirectory=$BK_DIR/journal
+export BK_ledgerDirectories=$BK_DIR/ledgers
+export BK_indexDirectories=$BK_DIR/index
+export BK_CLUSTER_ROOT_PATH=/${PRAVEGA_PATH}/${PRAVEGA_CLUSTER_NAME}/${BK_CLUSTER_NAME}
+
+export BK_tlsProvider=OpenSSL
+export BK_tlsKeyStoreType=JKS
+export BK_tlsKeyStore=/var/private/tls/bookie.keystore.jks
+export BK_tlsKeyStorePasswordPath=/var/private/tls/bookie.keystore.passwd
+export BK_tlsTrustStoreType=JKS
+export BK_tlsTrustStore=/var/private/tls/bookie.truststore.jks
+export BK_tlsTrustStorePasswordPath=/var/private/tls/bookie.truststore.passwd
 
 echo "wait for zookeeper"
-until ${ZK_HOME}/bin/zkCli.sh -server $ZK_URL ls /; do sleep 2; done
-
-echo "create the zk root"
-# Silence exit codes with "|| :" as the commands can safely fail with a "Node already exists" error
-${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH} || :
-${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH}/${PRAVEGA_CLUSTER_NAME} || :
-${ZK_HOME}/bin/zkCli.sh -server $ZK_URL create /${PRAVEGA_PATH}/${PRAVEGA_CLUSTER_NAME}/${BK_CLUSTER_NAME} || :
+until zk-shell --run-once "ls /" ${ZK_URL}; do sleep 5; done
 
-echo "format the bookie"
-# Silence exit codes with "|| :" as the command can safely fail if another instance has already formatted the bookie
-BOOKIE_CONF=${BK_HOME}/conf/bk_server.conf ${BK_HOME}/bin/bookkeeper shell metaformat -nonInteractive || :
+echo "delete bookie cookie"
+sed -i "s|metadataServiceUri=.*\$|metadataServiceUri=${BK_metadataServiceUri}|" /opt/bookkeeper/conf/bk_server.conf
+/opt/bookkeeper/bin/bookkeeper shell bookieformat -nonInteractive -force -deleteCookie
```

Comment:
I am not sure about this step, mayve I missing bit of context
why are you always clearing the bookie ?

second point: where is the 'metaformat' ? which is now called 'initnewcluster' ?

Reply:
@eolivelli correct me if I'm wrong, but I don't think we need to init a new cluster every time an instance of BK is started. I've added the `bookieformat` to delete the cookie related to that Bookie instance ID. When an instance running on a node was deleted and then recreated, we were getting an invalid cookie exception. The `bookieformat` attempts to fix that.

Let me know if we are following the right approach.

Reply:
To confirm, is this creating a new bookie from scratch? if so, it is correct to reformat the disk, and otherwise it isn't. See the documentation:

https://bookkeeper.apache.org/docs/4.7.3/admin/bookies/

But assuming that we are starting a bookie from scratch, then why are we finding these cookies in the first place?

Reply:
> I am not sure about this step, mayve I missing bit of context
> why are you always clearing the bookie ?

@eolivelli correct me if I'm misunderstanding something... I'm clearing the bookie because of the following use case: say that we have a bookie running in a node, then we scale down the BK cluster and that bookie and all its associated volumes are deleted, then we scale up the BK again and a new bookie is created in the same node, having the same instance ID. If we don't format the bookie metadata, the bookie will find existing metadata from the previous bookie and it will fail to start and throw a `InvalidCookieException` exception.

> second point: where is the 'metaformat' ? which is now called 'initnewcluster' ?

`initnewcluster` is handled by the official BK Docker scripts here: https://github.com/apache/bookkeeper/blob/release-4.7.3/docker/scripts/init_bookie.sh#L47. So that script will take care of initializing a new cluster if there isn't any available. 

> To confirm, is this creating a new bookie from scratch? if so, it is correct to reformat the disk, and otherwise it isn't.

@fpj yes, every time a Pravega BK container is started, it's assuming it's a new bookie.

> But assuming that we are starting a bookie from scratch, then why are we finding these cookies in the first place?

Because a previous Bookie with the same instance ID could have been created and destroyed before, leaving conflicting metadata behind that would cause the Bookie to fail when starting.